onadata/apps/logger/data_migration/restore_backup.py

```diff
@@ -19,33 +31,82 @@ def copy_attrs(from_obj, to_obj, attrs):
+        )
+        # XXX: Temporarily give form backup old xml so that xml comparison in
+        # data migration works as exected. Otherwise it would compare against
+        # the same xmls
+        self.xform_backup.xml = self.old_xml
         data_migrator_factory(
             self.xform_backup, self.xform,
-            backup_data=False, **reversed_decisions
+            backup_data=False, **decisions
         ).migrate()
 
+    @classmethod
+    def _merge_migration_changes(cls, changes_list):
+        empty_changes = MigrationDecisioner.construct_changes()
+        return cls._merge_migration_changes_aux(changes_list, empty_changes)
+
+    @classmethod
+    def _merge_migration_changes_aux(cls, changes_list, merged_changes):
+        if not len(changes_list):
+            return merged_changes
+        changes = changes_list.pop(0)
+        new_merged_changes = cls._merge_changes(merged_changes, changes)
+        return cls._merge_migration_changes_aux(changes_list, new_merged_changes)
+
+    @classmethod
+    def _merge_changes(cls, curr_changes, new_changes):
+        merge_and_remove_duplicates = lambda x1, x2: list(set(x1) | set(x2))
+
+        return MigrationDecisioner.construct_changes(
+            new=merge_and_remove_duplicates(curr_changes[NEW_FIELDS_KEY],
+                                            new_changes[NEW_FIELDS_KEY]),
+            removed=merge_and_remove_duplicates(curr_changes[RM_FIELDS_KEY],
+                                                new_changes[RM_FIELDS_KEY]),
+            modified=cls._transitive_merge(curr_changes[MOD_FIELDS_KEY],
+                                           new_changes[MOD_FIELDS_KEY]),
+        )
+
     @staticmethod
-    def get_reversed_migration_decisions(changes):
-        reversed_changes = MigrationDecisioner.reverse_changes(changes)
-        return MigrationDecisioner.convert_changes_to_decisions(reversed_changes)
+    def _transitive_merge(curr_changes, new_changes):
+        curr_changes = {
+            new_var_name: new_changes.pop(prev_var_name, prev_var_name)
+            for new_var_name, prev_var_name in curr_changes.items()
+        }
+        curr_changes.update(new_changes)
+        return curr_changes
 
-    def _get_xform_backup(self, version, restore_last):
+    @staticmethod
+    def _get_path_in_version_tree(xform, backup):
+        return VersionTree.objects.find_path(xform.version_tree, backup.version_tree)
+
+    @classmethod
+    def _get_migration_changes_in_between(cls, xform, backup):
+        path_up, path_down = cls._get_path_in_version_tree(xform, backup)
+        path_up_changes = map(cls._rev_changes, cls._get_changes_from_vt_path(path_up))
+        path_down_changes = cls._get_changes_from_vt_path(path_down)
+        return path_up_changes + path_down_changes
+
+    @staticmethod
+    def _get_changes_from_vt_path(path):
+        return map(lambda vt: vt.version.migration_changes, path)
+
+    @staticmethod
+    def _rev_changes(changes):
```

Comment:
I don't think _rev_changes is useful, just using MigrationDecisioner.reverse_changes in one place it is used looks fine. I don't see any gain in terms of readability here.

Reply:
Yeah, it's useless now, should have spotted that. At early stage of the code i used to use this method in more than one place. Fixe now.